ballerina/message_store.bal

```diff
@@ -0,0 +1,162 @@
+    # Enable or disable the queue declaration. Defaults to true
+    boolean enabled = true;
+    # Additional queue configuration to match the RabbitMQ queue declaration
+    QueueConfig queueConfig?;
+|};
+
+# Represents a RabbitMQ store client configuration for publishing messages.
+public type StoreClientPublishConfiguration record {|
+    # The RabbitMQ exchange to publish messages to. Defaults to an empty string
+    string exchange = "";
+    # The RabbitMQ delivery tag for the message. Optional
+    int deliveryTag?;
+    # The RabbitMQ message properties. Optional
+    BasicProperties properties?;
+|};
+
+# Represents a RabbitMQ message store implementation.
+public isolated client class MessageStore {
+    *messaging:Store;
+
+    private final Client rabbitMqClient;
+    private final readonly & StoreClientPublishConfiguration publishConfig;
+    private final string queueName;
+    private map<int?> consumedMessageDeliverTags;
+    private final readonly & StoreClientConfiguration clientConfig;
+
+    # Initializes a new instance of the RabbitMqStore class.
+    #
+    # + queueName - The name of the RabbitMQ queue to use for storing messages
+    # + clientConfig - The RabbitMQ client configuration
+    public isolated function init(string queueName, *StoreClientConfiguration clientConfig) returns error? {
+        self.clientConfig = clientConfig.cloneReadOnly();
+        self.rabbitMqClient = check new (clientConfig.host, clientConfig.port, clientConfig.connectionData);
+        self.publishConfig = clientConfig.publishConfig.cloneReadOnly();
+        self.queueName = queueName;
+        self.consumedMessageDeliverTags = {};
+
+        if clientConfig.declareQueue.enabled {
+            check self.rabbitMqClient->queueDeclare(queueName, clientConfig.declareQueue.queueConfig);
+        }
+    }
+
+    # Stores a message in the RabbitMQ message store.
+    #
+    # + payload - The message payload to be stored
+    # + return - An error if the message could not be stored, or `()`
+    isolated remote function store(anydata payload) returns error? {
+        error? result = self.rabbitMqClient->publishMessage({
+            content: payload,
+            routingKey: self.queueName,
+            ...self.publishConfig
+        });
+
+        if result is error {
+            return error("Failed to store message in RabbitMQ", cause = result);
+        }
+    }
+
+    # Retrieves the top message from the RabbitMQ message store without removing it. Retrieving
+    # concurrently without acknowledgment will return the next message in the queue if the previous one
+    # is not acknowledged.
+    #
+    # + return - The retrieved message, or `()` if the store is empty, or an error if an error occurs
+    isolated remote function retrieve() returns messaging:Message|error? {
+        lock {
+            AnydataMessage|error message = self.rabbitMqClient->consumeMessage(self.queueName, false);
+            if message is error {
```

Comment:
If messages are not found, should not the client return a nil value ?
```suggestion
            AnydataMessage|error? message = self.rabbitMqClient->consumeMessage(self.queueName, false);
            if message is () {
                // no message found 
            }
```

Reply:
Ideally `consumeMessage` API should have returned AnydataMessage (or subtype), error, or nil. I understand this is a breaking change, but should not we do that as well ?

Reply:
Since it is breaking change(all `consumeMessage` calls will break), I did not add that. @ThisaruGuruge Appreciate your input on this

Reply:
@aashikam Is there any specific reason for not returning a nil value here?

Reply:
![931ebc76544da004e58b6679e4051ac606614b0a_hq](https://github.com/user-attachments/assets/b946ae91-b8f6-4cb4-ac50-66b894a7c7d1)

There was no specific reason, but to be consistent across messaging modules similar to Kafka: https://github.com/ballerina-platform/ballerina-library/issues/2818#issuecomment-1091443625

Give me something for finding this.

Reply:
Should've been a part of this https://docs.google.com/document/d/1O_RRExICxYxzBoOpmxayqBRuf_UGzF-4BxVuQ_M7MI8/edit?usp=sharing,

Reply:
Let's keep this as it is for now since this will be a major version change.

Reply:
No the idea is if you use `consumePayload` that will return ().